### PR TITLE
ROS2 migration (foxy)

### PR DESCRIPTION
**Pull request by hangst**
Ported according to changes in `filters::FilterBase` and `filters::FilterChain`, with working examples and tests. 

```
[==========] Running 6 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 6 tests from ScanToScanFilterChain
[ RUN      ] ScanToScanFilterChain.BadConfiguration
[       OK ] ScanToScanFilterChain.BadConfiguration (13 ms)
[ RUN      ] ScanToScanFilterChain.IntensityFilter
********** SCAN **********
0 - 2000: 10
2000 - 4000: 0
4000 - 6000: 0
6000 - 8000: 0
8000 - 10000: 0
10000 - 12000: 0
12000 - 14000: 0
14000 - 16000: 0
16000 - 18000: 0
18000 - 20000: 0
20000 - 22000: 0
22000 - 24000: 0
24000 - 26000: 0
26000 - 28000: 0
28000 - 30000: 0
30000 - 32000: 0
32000 - 34000: 0
34000 - 36000: 0
36000 - 38000: 0
38000 - 40000: 0
40000 - 42000: 0
42000 - 44000: 0
44000 - 46000: 0
46000 - 48000: 0
[       OK ] ScanToScanFilterChain.IntensityFilter (6 ms)
[ RUN      ] ScanToScanFilterChain.InterpFilter
[       OK ] ScanToScanFilterChain.InterpFilter (6 ms)
[ RUN      ] ScanToScanFilterChain.ShadowFilter
[       OK ] ScanToScanFilterChain.ShadowFilter (5 ms)
[ RUN      ] ScanToScanFilterChain.ArrayFilter
[INFO] [1604546752.158335593] [array_filter_chain]: LaserArrayFilter cleaning and reallocating due to larger scan size
[       OK ] ScanToScanFilterChain.ArrayFilter (7 ms)
[ RUN      ] ScanToScanFilterChain.MaskFilter
[INFO] [1604546752.163838103] [mask_filter_chain]: LaserScanMaskFilter: laser: 2 directions will be masked.
[       OK ] ScanToScanFilterChain.MaskFilter (3 ms)
[----------] 6 tests from ScanToScanFilterChain (40 ms total)

[----------] Global test environment tear-down
[==========] 6 tests from 1 test case ran. (40 ms total)
[  PASSED  ] 6 tests.
```

### CODE REVIEW

**Comment by jonbinney** [comment]
Whoa thats a lot of changes. Thanks for doing this! I haven't been following the ros2 branch super closely - what ros2 distro was it working on before?

**Comment by hangst** [comment]
The last commits were 2+ years ago, so I guess it was targeted at Ardent? I couldn't get it work with Dashing and Foxy.

File: CMakeLists.txt [hunk 1]
```diff
@@ -5,113 +5,49 @@ project(laser_filters)
...
-  ${message_filters_LIBRARIES}
-  ${tf2_ros_LIBRARIES}
-  ${laser_geometry_LIBRARIES}
-  ${angles_LIBRARIES}
-  ${rostime_LIBRARIES}
-  ${Boost_LIBRARIES}
-)
+find_package(ament_cmake_auto REQUIRED)
+ament_auto_find_build_dependencies()
+add_compile_options(-Wall -Werror)
```
Comment (jonbinney) [inline CMakeLists.txt]: Are `-Wall` and `-Werror` recommended for ROS2 packages, or did you add these to help with debugging?

File: CMakeLists.txt [hunk 2]
```diff
@@ -5,113 +5,49 @@ project(laser_filters)
...
-install(DIRECTORY include/${PROJECT_NAME}/
-  DESTINATION include)
+##############################################################################
+# Test
+##############################################################################
 
-install(FILES laser_filters_plugins.xml
-    DESTINATION share
-)
+if(BUILD_TESTING)
```
Comment (jonbinney) [inline CMakeLists.txt:38]: Thanks for updating the tests for ros2!

File: CMakeLists.txt (hunk 1 above)
Comment (hangst) [reply CMakeLists.txt]: Ahh my apologies, yes I added them just for debugging.

**Comment by jonbinney** [comment]
@hangst sorry I forgot about this for a bit; I've merged it. Thanks!

**Comment by hangst** [comment]
Great, thanks!

**Pull request merged**